### PR TITLE
[Core] Improve  performance for parsing RFC1123 format string

### PR DESCRIPTION
**Pull request by mssfang**
Fixes: https://github.com/Azure/azure-sdk-for-java/issues/22000

Thanks to Jonathon's prototype: https://github.com/JonathanGiles/RFC1123-parser/blob/master/src/main/java/net/jonathangiles/tools/rfc1123/RFC1123Parser.java

Performance for parsing the RFC1123 format string:

- JDK takes                                  ~.5ms/op, 
- Jonathon's prototype takes      ~.02ms/op

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java [hunk 1]
```diff
@@ -3,15 +3,21 @@
...
 import java.time.ZoneId;
+import java.time.ZoneOffset;
 import java.time.format.DateTimeFormatter;
 import java.util.Locale;
 
 /**
  * Wrapper over java.time.OffsetDateTime used for specifying RFC1123 format during serialization and deserialization.
  */
 public final class DateTimeRfc1123 {
+    private final ClientLogger logger = new ClientLogger(DateTimeRfc1123.class);
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java:19]: Given the performance implications of this class (highly instantiated), I think we should use a static `ClientLogger`

File: sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java [hunk 2]
```diff
@@ -46,6 +52,75 @@ public OffsetDateTime getDateTime() {
         return this.dateTime;
     }
 
+    /**
+     * Parses the RFC1123 format datetime string into OffsetDateTime.
+     *
+     * @param date The datetime string in RFC1123 format
+     * @return The underlying OffsetDateTime.
+     */
+    private OffsetDateTime parse(final String date) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java:61]: nit: `static`, and same for the methods this is calling

File: sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java [hunk 3]
```diff
@@ -46,6 +52,75 @@ public OffsetDateTime getDateTime() {
...
+        return OffsetDateTime.of(
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java:62]: Have an early out for if the `date` string is less than 25 characters

File: sdk/core/azure-core/src/test/java/com/azure/core/util/DateTimeRfc1123Tests.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Tests {@link DateTimeRfc1123}.
+ */
+public class DateTimeRfc1123Tests {
+    private static final DateTimeFormatter RFC1123_DATE_TIME_FORMATTER =
+        DateTimeFormatter.ofPattern("EEE, dd MMM yyyy HH:mm:ss 'GMT'").withZone(ZoneId.of("UTC")).withLocale(Locale.US);
+
+    @Test
+    public void parseDateTimeRfc1123String() {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/test/java/com/azure/core/util/DateTimeRfc1123Tests.java:25]: Let's add a few more tests than this

File: sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java [hunk 5]
```diff
@@ -35,7 +41,7 @@ public DateTimeRfc1123(OffsetDateTime dateTime) {
      * @param formattedString The datetime string in RFC1123 format
      */
     public DateTimeRfc1123(String formattedString) {
-        this.dateTime = OffsetDateTime.parse(formattedString, DateTimeFormatter.RFC_1123_DATE_TIME);
+        this.dateTime = parse(formattedString);
```
[the remark this answers is not in the document]
Comment (JonathanGiles) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java:44]: This was my expectation. We should fall back to the code that we know works.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java (hunk 3 above)
Comment (mssfang) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java]: what about let the offsetDateTime.of() to check it.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java:45]: Added the fallback to OffsetDateTime's parse method in parse() now

**Review by alzimmermsft — APPROVED** [review]

**Review by JonathanGiles — APPROVED** [review]

**Pull request merged**